Durability.md

```diff
@@ -0,0 +1,218 @@
+do we ensure that?
+
+1. Use etcd to grant a lease on a key. The primary node is only allowed to operate as primary
+   when it's holding a valid lease. If the primary node dies, the lease expires after a timeout
+   period, and a new node is allowed to become the primary.
+
+2. Use S3 to store the lease. S3's consistency guarantees are more lenient, so in theory you
+   cannot do this safely. In practice, it would probably be OK if you make the lease times and
+   timeouts long enough. This has the advantage that we don't need to introduce a new
+   component to the architecture.
+
+3. Use Raft or Paxos, with the WAL safekeepers acting as the Acceptors to form the quorum. The
+   next chapter describes this option.
+
+
+Built-in Paxos
+--------------
+
+The WAL safekeepers act as PAXOS Acceptors, and the Processing nodes
+as both Proposers and Learners.
+
+Each WAL safekeeper holds an Epoch value in addition to the VCL and
+the WAL. Each request by the primary to safekeep WAL is accompanied by
+an Epoch value. If a safekeeper receives a request with Epoch that
+doesn't match its current Accepted Epoch, it must ignore (NACK) it.
+(In different Paxos papers, Epochs are called "terms" or "round
+numbers")
+
+When a node wants to become the primary, it generates a new Epoch
+value that is higher than any previously observed Epoch value, and
+globally unique.
+
+
+Accepted Epoch: 555                VCL                   LSN
+                                     |                     |
+                                     V                     V
+.................ccccccccccccccccccccXXXXXXXXXXXXXXXXXXXXXXX
+Archived WAL       Completed WAL          In-flight WAL
+
+
+Primary node startup:
+
+1. Contact all WAL safekeepers that you can reach (if you cannot
+   connect to a quorum of them, you can give up immediately). Find the
+   latest Epoch among them.
+
+2. Generate a new globally unique Epoch, greater than the latest Epoch
+   found in previous step.
+
+2. Send the new Epoch in a Prepare message to a quorum of
+   safekeepers. (PAXOS Prepare message)
+
+3. Each safekeeper responds with a Promise. If a safekeeper has
+   already made a promise with a higher Epoch, it doesn't respond (or
+   responds with a NACK). After making a promise, the safekeeper stops
+   responding to any write requests with earlier Epoch.
+
+4. Once you have received a majority of promises, you know that the
+   VCL cannot advance on the old Epoch anymore. This effectively kills
+   any old primary server.
+
+5. Find the highest written LSN among the quorum of safekeepers (these
+   can be included in the Promise messages already). This is the new
+   VCL.  If a new node starts the election process after this point,
+   it will compute the same or higher VCL.
+
+6. Copy the WAL from the safekeeper with the highest LSN to the other
+   safekeepers in the quorum, using the new Epoch. (PAXOS Accept
+   phase)
```

Comment:
👍 yeah, I also was thinking about that place. Raft does this differently in a way that we can't support since processing nodes are not proposing themselves. So following paxos here is better.

Reply:
> in a way that we can't support since processing nodes are not proposing themselves

But why? It would be fine to allow VCL advancement by the new primary only after old missing WAL (preserving the epoch) copied on majority + a (dummy) entry from the new epoch is also written on it, like Raft does. This choice is argued in 5.4.2 section of
https://raft.github.io/raft.pdf

Restamping records with the new epoch instead is possible, but sounds more complicated and expensive; note that you must renumber them *everywhere*, including all servers who already have the record.

Reply:
Hmm. I don't think we need to stamp the records with epoch at all. Each WAL safekeeper has a *current* epoch value, and the WAL. The current epoch is just a variable, you don't need to write it in the log, or associate it with any particular WAL records.

Reply:
I think what I'm proposing is Multi-Paxos: https://en.wikipedia.org/wiki/Paxos_(computer_science)#Multi-Paxos. Mapping the terms:

The Processing node is a Client, Proposer and Learner. The WAL safekeepers are Acceptors.

When the processing node starts up and wants to become the primary, it initiates the elections, and perform the full Paxos protocol. This is the "Multi-Paxos without failures" sequence. Subsequent writes to the WAL can skip the phase 1, following the "Multi-Paxos when phase 1 can be skipped" sequence.

Reply:
> This choice is argued in 5.4.2 section of
Yeah, I was talking about that election restriction in 5.4.1: "Raft uses the voting process to prevent a candidate from winning an election unless its log contains all committed entries". If we assume that processing node is stateless it have no logs but we allow it to be elected and there is nothing to replicate in previous term. So we need to brew something different: either some recovery mechanism through master or allow safe keepers to communicate between each other.

> I don't think we need to stamp the records with epoch at all.
Are you talking about network or disk? In network all records should be packed within some container stamped with epoch so that safe keeper can distinguish between records from competing masters. With disk I can't came up right away with example when we need to know term for each record -- I'll think more.

Reply:
> > This choice is argued in 5.4.2 section of
> 
> Yeah, I was talking about that election restriction in 5.4.1: "Raft uses the voting process to prevent a candidate from winning an election unless its log contains all committed entries". If we assume that processing node is stateless it have no logs but we allow it to be elected and there is nothing to replicate in previous term. So we need to brew something different: either some recovery mechanism through master or allow safe keepers to communicate between each other.

I'm imagining that the processing node is the Client, not a Server. The primary processing node connects to the leader WAL safekeeper. Yeah, the safekeepers will need to communicate with each other in Raft.

> > I don't think we need to stamp the records with epoch at all.
> 
> Are you talking about network or disk? In network all records should be packed within some container stamped with epoch so that safe keeper can distinguish between records from competing masters. With disk I can't came up right away with example when we need to know term for each record -- I'll think more.

On disk. You're right that you need the epoch in the network requests. Or you rely on the leader safekeeper to only allow one primary processing node to be connected at a time, and when you connect you first ask the safekeeper what the current end-of-WAL is, before you start to append.

Reply:
> I'm imagining that the processing node is the Client, not a Server. 

Wouldn't that lead to same race between two competing clients?

I was imagining processing node being a Server which can propose himself, but can't vote (so destined to be leader) and safe keeper being a Server which can vote but can't propose (forever follower). Also in this setup we don't know amount of participants from the start: set of safekeepers is fixed but set of processing is potentially unbounded.
So more or less same setup as you proposed with Paxos.